engine/variableValue.h

```diff
@@ -139,19 +141,28 @@ namespace minsky
       setXVector(std::move(xv));
       return d;
     }
-    size_t numElements() const {
+    // For feature 47
+    std::vector<size_t> index;
+    
+    double sparsityRatio=0.0;       
+    
+    size_t numDenseElements() const {                                                
       size_t s=1;
       for (auto& i: xVector) s*=i.size();
       return s;
     }
-
-    template <class T>
-    void setXVector_(T x) {
-      size_t prevNumElems=numElements();
-      m_xVector=x;
-      if (idx()==-1 || prevNumElems<numElements())
-        allocValue();
-    }
+    
+    size_t numSparseElements = 0.0;
+    
+    template <class T>                                            
+    void setXVector_(T x) {    
+      size_t prevNumElems;
+      if (sparsityRatio <= 0.5) prevNumElems=numDenseElements();
+      else prevNumElems=numSparseElements;    
+      m_xVector=x;    
+      if (idx()==-1 || (prevNumElems<numDenseElements() || prevNumElems<numSparseElements))    
+        allocValue();    
```

Comment:
I'm not sure about this. Now that the number of elements allocated is uncoupled from xVector, it may no longer make sense allocValue here. Alternatively, we assume that the index vector has been set correctly. This entails that the index vector needs to be encapsulated in the same way as xvector, such that setting the index vector also forces an allocValue if the index vector changes length.

Reply:
Would an appropriately defined template<> to set the index vector do the job?

template<T>
void setDataSize(T t) {
}

with a protected member m_dataSize set equal to t=index.size() or numDenseElement.size()?

Reply:
Probably better to have a setIndex() setter, and a index() getter that returns a const reference to the m_index member variable. That way, the allocated vector is the correct length.

Reply:
Does the following look okay (in variableValue.h)?

  class VariableValue: public VariableType, public XVectorMixin
  {
    CLASSDESC_ACCESS(VariableValue);
  private:
    std::vector<size_t> m_index;

  public:
    std::vector<size_t> getIndex() **const** { return m_index;}
    **const** void setIndex(std::vector<size_t> index) {	m_index=index;}

    size_t dataSize() **const** {
		return getIndex().size() ? getIndex().size(): numDenseElements();
	}    

I've put the "const" in boldface because my studies of C++ have still not got to the point where I can understand the meaning of constify in different positions in member declarations.

Reply:
In the first case, you're declaring the return type const (const void - which doesn't make much sense as you cannot mutate void anyway - but maybe the compiler allows it). In the second case, you're declaring the method const, which means you can call it on a const object.

WRT types, you can place the const before or after the type being qualified. I prefer before, because I speak English, and const is an adjective and that's where English puts adjectives. Other people seem to have the opposite choice. Finally, beloved of C++ job interviews, not so useful in real life,
T * const   -  the pointer is const, but the target is not
const T* const - both the pointer and target is const
T const * const  - ditto.

Reply:
> In the first case, you're declaring the return type const (const void - which doesn't make much sense as you cannot mutate void anyway - but maybe the compiler allows it).

Thanks, now I understand this. I didn't know the const could go before or after the type. I've mostly seen it after in type declarations, but it makes sense that it go first because it is akin to an adjective.

Indeed, the compiler allows the second case with or without the const on the type. It does not allow the const on the method, however. I get the error:

error: passing ‘const std::vector<long unsigned int>’ as ‘this’ argument discards qualifiers [-fpermissive]
   m_index=index;

I guess this is because I modify m_index, so the object is not const.

> Finally, beloved of C++ job interviews, not so useful in real life,
T * const - the pointer is const, but the target is not
const T* const - both the pointer and target is const
T const * const - ditto.

Thanks, very useful advice for interviews, on this, as well as the linked lists previously :)